### PR TITLE
rpc|test:  Introduce "mnauth" RPC command to override masternode authentications

### PR DESCRIPTION
**Pull request by xdustinface**
Allows to fake verified MN P2P connections. Something i need for an other test. If someone has a better idea how to archive this, let me know :)

Based on #3929

### CODE REVIEW

File: test/functional/rpc_mnauth.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# Copyright (c) 2018-2020 The Dash Core developers
```
Comment (PastaPastaPasta) [inline test/functional/rpc_mnauth.py:2]:
```suggestion
# Copyright (c) 2021 The Dash Core developers
```

**Comment by UdjinM6** [comment]
Code looks good but... why?

> Something i need for another test.

Can you expand on this a bit pls?

**Comment by xdustinface** [comment]
> Code looks good but... why?
> 
> > Something i need for another test.
> 
> Can you expand on this a bit pls?

Im working on the `QGETDATA`/`QDATA` stuff which _should_ be restricted to be used by masternodes only. At least that was the initial proposed idea from the document and normal nodes should anyway not care about it. And while i was writing tests i realised its not possible to write direct P2P tests if there is such a restriction. Soo, i see/saw the following options:

1. Not restrict it to masternodes only at all 
2. Not restrict it to masternodes only for `-regtest`
3. Not restrict it to masternodes only if some command line parameter is set
4. Just fake MNAUTH for the testing P2P connection

I mean, actually i like having such a restriction but we could go with 1. if thats not a real concern? I guess the data should be cached then but it would make testing easier. Anyway, i don't like 2. and 3. because it wouldn't be a "full" test thats why i then took 4.

**Comment by UdjinM6** [comment]
Ok, so you want to use `mininode` to send/receive these new messages but to do this you want `dashd` you are connected to to think that this `mininode` is a masternode, right? Can't you just reuse `qwatch` for this?

**Comment by xdustinface** [comment]
>  Can't you just reuse `qwatch` instead?

Hm, not really. Maybe i just don't get how you think that would help but what i basically need is that the connected node has a verified protx hash assigned. This branch is still WIP but See https://github.com/xdustinface/dash/commit/aea5dd77b97c8e151b19497ccef9e60299975afe, https://github.com/xdustinface/dash/commit/03e9edb604bcc549017a9ea66249b06c4942817c. Its already implemented with the fake mnauth in https://github.com/xdustinface/dash/commit/c287a462b7ea60ee809bb2129da17943c3b3186d but im totally open to drop the implementation and this PR if you come up with something else which makes more sense, let me know your thoughts :)

**Comment by UdjinM6** [comment]
How about https://github.com/xdustinface/dash/commit/aea5dd77b97c8e151b19497ccef9e60299975afe#r45968519 ?

**Comment by xdustinface** [comment]
@UdjinM6 Got it now with `qwatch`! I like the idea, see a55ac9a where i implemented it :) Buut, what do you think about adding both? I mean, the `mnauth` RPC command is basically useful for tests in general (depending on what we might do in the future with the `MNAUTH` for sure 🙈), no? And it at least helps in #3953 to somewhat test independent connections like here https://github.com/dashpay/dash/pull/3953/files#diff-2829e166b218f8d2acb7af7fd4e966be5cd8a21c9cc97f13fc29a8fa4a940dc6R274-R303. It's obviously not super important thing to test but it's one more 😄

File: src/rpc/misc.cpp [hunk 2]
```diff
@@ -421,6 +422,42 @@ UniValue setmocktime(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
+UniValue mnauth(const JSONRPCRequest& request)
+{
+    if (request.fHelp || (request.params.size() != 3))
+        throw std::runtime_error(
+            "mnauth nodeId \"proTxHash\" \"publicKey\"\n"
+            "\nFake a MNAUTH for a connected node in regression test (-regtest only).\n"
```
Comment (UdjinM6) [inline src/rpc/misc.cpp:430]: nit: How about smth like
```suggestion
            "\nOverride MNAUTH processing results for the specified node with a user provided data (-regtest only).\n"
```
?

**Review by UdjinM6 — DISMISSED** [review]
utACK

File: test/functional/rpc_mnauth.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        assert_raises_rpc_error(-8, "proTxHash invalid", masternode.node.mnauth,
+                                                         node_id,
+                                                         null_hash,
+                                                         public_key)
+        assert_raises_rpc_error(-8, "publicKey invalid", masternode.node.mnauth,
+                                                         node_id,
+                                                         protx_hash,
+                                                         null_hash)
+        assert(not masternode.node.mnauth(-1, protx_hash, public_key))
+
```
Comment (PastaPastaPasta) [inline test/functional/rpc_mnauth.py:54]:
```suggestion


```

**Review by PastaPastaPasta — APPROVED** [review]
one nit, otherwise utACK

**Review by UdjinM6 — APPROVED** [review]
re-utACK

(tweaked PR title a bit)

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**